### PR TITLE
First step to rearrange files in tools folder

### PR DESCRIPTION
**Pull request by walterddr**
Changes including:
- introduced `linter/`, `testing/`, `stats/` folders in `tools/`
- move appropriate scripts into these folders
- change grepped references in the pytorch/pytorch repo

Next step
- introduce `build/` folder for build scripts
- moves .circleci/scripts and .jenkins/scripts that are not platform dependent to tools/

Test Plan
- CI (this is important b/c pytorch/test-infra also rely on some script reference, see: https://github.com/pytorch/test-infra/pull/44
- Aded tools/tests/

### CODE REVIEW

**Review by seemethere — APPROVED** [review]

File: mypy.ini [hunk 1]
```diff
@@ -38,8 +38,8 @@ files =
     test/test_type_hints.py,
     test/test_type_info.py,
     test/test_utils.py,
-    tools/clang_format_utils.py,
-    tools/clang_tidy.py,
+    tools/linter/clang_format_utils.py,
+    tools/linter/clang_tidy.py,
     tools/generate_torch_version.py,
     tools/render_junit.py,
     tools/stats_utils
```
Comment (walterddr) [inline mypy.ini:45]:
```suggestion
    tools/stats
```

File: tools/test/test_test_selection.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        'short_test2': 0.6,
+        'short_test3': 0.4,
+        'short_test4': 0.3,
+        'short_test5': 0.01,
+    }
+
+    def assert_shards_equal(
+        self, 
+        expected_shards: List[Tuple[float, List[str]]], 
+        actual_shards: List[Tuple[float, List[str]]]) -> None:
```
Comment (walterddr) [inline tools/test/test_test_selection.py:40]:
```suggestion
    def assert_shards_equal(
        self,
        expected_shards: List[Tuple[float, List[str]]],
        actual_shards: List[Tuple[float, List[str]]]
    ) -> None:
```

File: tools/stats/print_test_stats.py [hunk 3]
```diff
@@ -18,9 +18,9 @@
 
 import requests
 from typing_extensions import TypedDict
-from tools.stats_utils.s3_stat_parser import (newify_case, get_S3_object_from_bucket, get_test_stats_summaries_for_job,
-                                              Report, Status, Commit, HAVE_BOTO3, Version2Case, VersionedReport,
-                                              Version1Report, Version2Report, ReportMetaMeta)
+from tools.stats.s3_stat_parser import (newify_case, get_S3_object_from_bucket, get_test_stats_summaries_for_job,
+                                        Report, Status, Commit, HAVE_BOTO3, Version2Case, VersionedReport,
+                                        Version1Report, Version2Report, ReportMetaMeta)
```
Comment (samestep) [inline tools/stats/print_test_stats.py:23]:
```suggestion
from tools.stats.s3_stat_parser import (
    HAVE_BOTO3,
    Commit,
    Report,
    ReportMetaMeta,
    Status,
    Version1Report,
    Version2Case,
    Version2Report,
    VersionedReport,
    get_S3_object_from_bucket,
    get_test_stats_summaries_for_job,
    newify_case,
)
```

File: tools/stats/test_history.py [hunk 4]
```diff
@@ -7,8 +7,8 @@
 from signal import SIG_DFL, SIGPIPE, signal
 from typing import Dict, Iterator, List, Optional, Set, Tuple
 
-from tools.stats_utils.s3_stat_parser import (Report, get_cases,
-                                              get_test_stats_summaries)
+from tools.stats.s3_stat_parser import (Report, get_cases,
+                                        get_test_stats_summaries)
```
Comment (samestep) [inline tools/stats/test_history.py:11]:
```suggestion
from tools.stats.s3_stat_parser import (
    Report,
    get_cases,
    get_test_stats_summaries,
)
```

File: tools/test/test_stats.py [hunk 5]
```diff
@@ -2,11 +2,11 @@
...
-from tools import print_test_stats
-from tools.stats_utils.s3_stat_parser import (Commit, Report, ReportMetaMeta,
-                                              Status, Version1Case,
-                                              Version1Report, Version2Case,
-                                              Version2Report)
+from tools.stats import print_test_stats
+from tools.stats.s3_stat_parser import (Commit, Report, ReportMetaMeta,
+                                        Status, Version1Case,
+                                        Version1Report, Version2Case,
+                                        Version2Report)
```
Comment (samestep) [inline tools/test/test_stats.py:9]:
```suggestion
from tools.stats.s3_stat_parser import (
    Commit,
    Report,
    ReportMetaMeta,
    Status,
    Version1Case,
    Version1Report,
    Version2Case,
    Version2Report,
)
```

File: tools/README.md [hunk 6]
```diff
@@ -53,17 +53,17 @@ Developer tools which you might find useful:
...
-* [test_history.py](test_history.py) - Query S3 to display history of a single
+* [stats/test_history.py](stats/test_history.py) - Query S3 to display history of a single
   test across multiple jobs over time.
-* [trailing_newlines.py](trailing_newlines.py) - Take names of UTF-8 files from
+* [linter/trailing_newlines.py](linter/trailing_newlines.py) - Take names of UTF-8 files from
   stdin, print names of nonempty files whose contents don't end in exactly one
   trailing newline, exit with status 1 if no output printed or 0 if some
   filenames were printed.
-* [translate_annotations.py](translate_annotations.py) - Read [Flake8][] or
+* [linter/translate_annotations.py](linter/translate_annotations.py) - Read [Flake8][] or
```
Comment (samestep) [inline tools/README.md:66]: for consistency, the things that are now under `linter/` or `stats/` should be in sub-bullets (similar to `shared/`, `fast_nvcc/`, and `amd_build/`)

**Review by samestep — APPROVED** [review]
looks good, thank you! just a few cosmetic comments about `tools/README.md` and about `import`s (see #60119)

**Pull request closed without merging**